File: crates/klyron_utils/bindings/cpp/builder.cpp
#include "builder.hpp"

namespace klyron {

class Klyron::UtilsBuilder::Impl {
public:
  Klyron::UtilsConfig config;
  bool verbose = false;
};

Klyron::UtilsBuilder::Klyron::UtilsBuilder()
  : impl_(std::make_unique<Impl>()) {
}

Klyron::UtilsBuilder& Klyron::UtilsBuilder::withConfig(const Klyron::UtilsConfig& config) {
  impl_->config = config;
  return *this;
}

Klyron::UtilsBuilder& Klyron::UtilsBuilder::verbose(bool v) {
  impl_->verbose = v;
  return *this;
}

Klyron::UtilsBuilder::Klyron::UtilsInstance Klyron::UtilsBuilder::build() {
  Klyron::UtilsInstance inst;
  inst.config = impl_->config;
  inst.verbose = impl_->verbose;
  return inst;
}

} // namespace klyron

File: crates/klyron_utils/bindings/rust/builder.rs
//! Builder pattern for klyron_utils

use crate::types::Klyron::UtilsConfig;

#[derive(Debug, Default)]
pub struct Klyron::UtilsBuilder {
    config: Option<Klyron::UtilsConfig>,
    verbose: bool,
}

impl Klyron::UtilsBuilder {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_config(mut self, config: Klyron::UtilsConfig) -> Self {
        self.config = Some(config);
        self
    }

    pub fn verbose(mut self, verbose: bool) -> Self {
        self.verbose = verbose;
        self
    }

    pub fn build(self) -> anyhow::Result<Klyron::UtilsInstance> {
        let config = self.config.unwrap_or_default();
        Ok(Klyron::UtilsInstance { config, verbose: self.verbose })
    }
}

#[derive(Debug)]
pub struct Klyron::UtilsInstance {
    pub config: Klyron::UtilsConfig,
    pub verbose: bool,
}
